sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,250 @@
+import numpy as np
+
+from ..base import BaseEstimator, clone
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if estimator is None:
+        raise ValueError("base_estimator cannot be None")
+
+    if not hasattr(estimator, "predict_proba"):
+        raise ValueError("base_estimator should implement predict_proba!")
+
+
+class SelfTrainingClassifier(BaseEstimator):
+    """Self-training classifier
+
+    Parameters
+    ----------
+    base_estimator : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``self.base_estimator_`` attribute.
+
+    threshold : float
+        Threshold above which predictions are added to the labeled dataset.
+        Should be in [0, 1).
+
+    max_iter : integer
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0.
+
+    Attributes
+    ----------
+    base_estimator_ : estimator object
+        The fitted estimator.
+
+    y_labeled_ : array, shape = (n_samples,)
+        The labels used for the final fit of the classifier.
+
+    y_labeled_iter_ : array, shape = (n_samples,)
+        The iteration in which each sample was labeled. When a sample has
+        iteration 0, the sample was already labeled in the given dataset. When
+        a sample has iteration -1, the sample was not labeled in any iteration.
+
+    n_iter_ : int
+        The amount of iterations the classifier during fitting.
+
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import datasets
+    >>> from sklearn.semi_supervised import SelfTrainingClassifier
+    >>> from sklearn.svm import SVC
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTrainingClassifier(svc)
+    >>> iris = datasets.load_iris()
+    >>> rng = np.random.RandomState(42)
+    >>> random_unlabeled_points = rng.rand(len(iris.target)) < 0.3
+    >>> labels = np.copy(iris.target)
+    >>> labels[random_unlabeled_points] = -1
+    >>> self_training_model.fit(iris.data, labels)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
```

Comment:
You don't need to make a copy of iris.target, might as well save a line

Concerning the order I would first set the random state, then set up the datasets, and then define and use the classifiers but that's mostly personal taste.